### ISSUE TITLE
CassandraKVS executor thread pool size likely too small

### ISSUE THREAD

**Issue by jnelson15** [palantir/atlasdb#2524]
The CassandraKVS uses a fixed executor of size `configManager.getConfig().poolSize() * configManager.getConfig().servers().size()` to run tasks. This seems odd, though, since we could get a value greater than that requests to a single client pool (or small set of them), which would exhaust the underlying pool and cause blocking (until potential timeouts or task completion), and thus would prevent any other under-utilized client pools from being hit since the executor thread pool is empty. Furthermore, this is a bit over-restricting even when evenly distributed since the client pools can grow to `maxConnectionBurstSize` which is usually much bigger than `poolSize`. I'm a little hesitant to just jack up the executor size, though, since we're probably getting some implicit throttling from this and could cause timeouts on the underlying client pools when we get bursts of requests. Any thoughts?

**Comment by jeremyk-91** [comment]
Note: We still need to address this because of the subsequent reversion of #2534 in #2637

**Comment by hsaraogi** [comment]
@jnelson15 are you planning on working on this?

**Comment by jnelson15** [comment]
I can, but not totally sure what the better solution is. Simplest would be just increasing the size of the original fixed thread pool. Downside is we'll have a lot more threads sitting around for the lifetime of the server but maybe that's not a huge deal. What do you think?

**Comment by jnelson15** [comment]
Also doesn't completely solve the original problem. But per-host routing at this layer could get messy since the cass client pool stuff does failovers.

**Comment by fsamuel-bs** [comment]
As per the email thread, I guess our consensus was to have #2534 with a SynchronousQueue. @nziebart for SA.

**Comment by jnelson15** [comment]
Would we still do caller runs in that case? Otherwise it'll reject tasks once it hits the max, right?

**Comment by fsamuel-bs** [comment]
I think we'd just hit the policy if the queue rejects a job (i.e. returns `false` to an `offer` call). So I think that having the default policy (AbortPolicy) is fine here.

**Comment by jnelson15** [comment]
Yeah but in the original implementation we had an unbounded LinkedBlockingQueue so it would never reject operations. So bursty workflows could get new failures that otherwise would just block for a while. Not sure which is better but worth considering.
